File: verifiers/go/witness_cosignature.go
package verifier

import (
"crypto/ed25519"
"encoding/hex"
"encoding/json"
"fmt"
)

type witnessCosignature struct {
WitnessID    string `json:"witness_id"`
PublicKeyHex string `json:"public_key_hex"`
SignatureHex string `json:"signature_hex"`
}

type witnessEnvelope struct {
RootHash          string               `json:"root_hash"`
WitnessThreshold  int                  `json:"witness_threshold"`
WitnessCosignatures []witnessCosignature `json:"witness_cosignatures"`
}

func VerifyWitnessEnvelopeJSON(raw []byte) error {
var env witnessEnvelope
if err := json.Unmarshal(raw, &env); err != nil {
return err
}
root, err := hex.DecodeString(env.RootHash)
if err != nil || len(root) != 32 {
return fmt.Errorf("invalid root hash")
}
payload := append([]byte("OLY:WITNESS:V1|"), root...)
valid := map[string]struct{}{}
for _, c := range env.WitnessCosignatures {
pk, err := hex.DecodeString(c.PublicKeyHex)
if err != nil || len(pk) != ed25519.PublicKeySize {
continue
}
sig, err := hex.DecodeString(c.SignatureHex)
if err != nil || len(sig) != ed25519.SignatureSize {
continue
}
if ed25519.Verify(ed25519.PublicKey(pk), payload, sig) {
valid[c.WitnessID] = struct{}{}
}
}
if len(valid) < env.WitnessThreshold {
return fmt.Errorf("threshold not met")
}
return nil
}

File: verifiers/js/witness_cosignature.mjs
import { readFileSync } from 'node:fs';
import { verify } from 'node:crypto';

export function verifyWitnessEnvelope(path) {
  const env = JSON.parse(readFileSync(path, 'utf8'));
  const root = Buffer.from(env.root_hash, 'hex');
  if (root.length !== 32) {
    return false;
  }
  const payload = Buffer.concat([Buffer.from('OLY:WITNESS:V1|', 'utf8'), root]);
  const valid = new Set();
  for (const c of env.witness_cosignatures ?? []) {
    try {
      const spkiPrefix = Buffer.from('302a300506032b6570032100', 'hex');
      const publicKey = Buffer.concat([spkiPrefix, Buffer.from(c.public_key_hex, 'hex')]);
      const sig = Buffer.from(c.signature_hex, 'hex');
      if (verify(null, payload, { key: publicKey, format: 'der', type: 'spki' }, sig)) {
        valid.add(c.witness_id);
      }
    } catch {
      // ignore malformed signature rows in scaffold verifier
    }
  }
  return valid.size >= Number(env.witness_threshold ?? 2);
}

if (process.argv[1] && process.argv[1].endsWith('witness_cosignature.mjs')) {
  const ok = verifyWitnessEnvelope(process.argv[2]);
  process.exit(ok ? 0 : 1);
}
